test/libdnf/base/test_goal.cpp

```diff
@@ -52,20 +53,55 @@ void BaseGoalTest::test_install() {
 }
 
 void BaseGoalTest::test_install_from_cmdline() {
+    // this test covers an old dnf bug described in the following steps:
+    // * specify a commandline package to install
+    // * the specified package has the same NEVRA as an available package in a repo
+    // * the package query uses NEVRA instead of id and is resolved into the available rather than the specified package
+    // * -> an unexpected package is installed
+
+    // add a repo with package 'one-0:1-1.noarch'
+    add_repo_rpm("rpm-repo1");
+
+    // add 'one-0:1-1.noarch' package from the command-line
+    std::filesystem::path rpm_path = PROJECT_BINARY_DIR "/test/data/repos-rpm/rpm-repo1/noarch/one-1-1.noarch.rpm";
+    auto cmdline_pkg = sack->add_cmdline_package(rpm_path, false);
+
+    // check if the available package is there and has lower id than the command-line package
+    // a package with lower id would be normally preferred if the packages are identical
+    libdnf::rpm::SolvQuery query(&(base->get_rpm_solv_sack()));
+    query.ifilter_available().ifilter_nevra(libdnf::sack::QueryCmp::EQ, {"one-0:1-1.noarch"});
+
+    std::vector<libdnf::rpm::Package> query_packages;
+    for (auto i : query) {
+        query_packages.push_back(i);
+    }
+    // confirm that the second package is the command-line package
+    CPPUNIT_ASSERT_EQUAL(cmdline_pkg.get_id().id, query_packages[1].get_id().id);
+
+    // the available package has lower id than the command-line package
+    CPPUNIT_ASSERT(query_packages[0].get_id().id < query_packages[1].get_id().id);
+
+    // install the command-line package
     libdnf::Goal goal(base.get());
-    std::filesystem::path rpm_path = PROJECT_BINARY_DIR "/test/data/cmdline-rpms/noarch/cmdline-1.2-3.noarch.rpm";
-    auto cmd_pkg = sack->add_cmdline_package(rpm_path, false);
-    goal.add_rpm_install(cmd_pkg, true);
+    goal.add_rpm_install(cmdline_pkg, true);
+
+    // resolve the goal and read results
     goal.resolve(false);
     auto install_set = goal.list_rpm_installs();
     auto reinstall_set = goal.list_rpm_reinstalls();
     auto upgrade_set = goal.list_rpm_upgrades();
     auto downgrade_set = goal.list_rpm_downgrades();
     auto remove_set = goal.list_rpm_removes();
     auto obsoleted_set = goal.list_rpm_obsoleted();
-    CPPUNIT_ASSERT_EQUAL(1lu, install_set.size());
-    CPPUNIT_ASSERT_EQUAL(install_set[0].get_full_nevra(), std::string("cmdline-0:1.2-3.noarch"));
-    CPPUNIT_ASSERT_EQUAL(install_set[0].get_repo()->get_id(), std::string("@commandline"));
+
+    // check if we're getting an expected NEVRA
+    std::vector<std::string> expected = {"one-0:1-1.noarch"};
+    CPPUNIT_ASSERT_EQUAL(expected, to_vector(install_set));
+
+    // also check that the installed package is identical to the command-line package
+    CPPUNIT_ASSERT_EQUAL(cmdline_pkg.get_id().id, install_set[0].get_id().id);
+    CPPUNIT_ASSERT_EQUAL(std::string("@commandline"), install_set[0].get_repo()->get_id());
```

Comment:
I guess that we can do it by a better way. We can directly compare packages. The transformation of package_set to vector of strings is not a nice way.
```
CPPUNIT_ASSERT_EQUAL(1lu, install_set.size());
CPPUNIT_ASSERT_EQUAL(cmdline_pkg, install_set[0])
```

Reply:
Sometimes you want to work with indexes and that's not possible because we removed operator[]: https://github.com/rpm-software-management/libdnf/blob/dnf-5-devel/CHANGES-DNF-5.md

I'm not saying it's nice, but it's relatively easy to understand and that counts.

Reply:
@dmach not sure how your comment addresses @j-mracek's remark, which I think is a good one, under two assumptions:
- We have `operator==()` for the `Package` class, which we do
- In case of the assert failing, it can actually print the packages that aren't equal in a useful way; I haven't checked this